File: packages/contracts/src/types/application.ts
/**
 * ApplicationStatus represents the full set of states in the state machine.
 *
 * Safety invariant (initial release):
 *   When INITIAL_SUBMISSION_MODE=stop_before_submit, transitions to "submitted"
 *   are unconditionally blocked. The "approved" state always flows to
 *   "stopped_before_submit", never to "submitted".
 *
 *   The "submitted" value exists in the type for future releases only and
 *   must never be reachable in initial mode.
 */
export type ApplicationStatus =
  | "draft"
  | "matched"
  | "packet_draft"
  | "packet_ready"
  | "human_review"
  | "approved"
  | "stopped_before_submit"
  | "submitted"; // UNREACHABLE in initial release — future explicit gate only

export type EventTrigger = "system" | "human";

export interface ApplicationEvent {
  id: string;
  applicationId: string;
  fromStatus: ApplicationStatus | null;
  toStatus: ApplicationStatus;
  triggeredBy: EventTrigger;
  actorId: string | null;
  note: string | null;
  createdAt: string; // ISO 8601
}

/**
 * Keys stay wire-shaped (snake_case) on purpose: a fingerprint is an opaque
 * blob that must be echoed back to the API byte-identical when an assisted run
 * starts, so the dashboard client never rewrites its keys.
 */
export interface PacketFingerprint {
  profile_version: number;
  resume_version: number | null;
  cover_letter_version: number | null;
  answer_versions: Record<string, number>;
  job_snapshot_hash: string;
  packet_hash: string;
}

export interface Application {
  id: string;
  jobId: string;
  candidateProfileId: string;
  status: ApplicationStatus;
  packetFingerprint: PacketFingerprint | null;
  events: ApplicationEvent[];
  createdAt: string; // ISO 8601
  updatedAt: string; // ISO 8601
}

export interface ApplicationSummary {
  id: string;
  jobId: string;
  candidateProfileId: string;
  status: ApplicationStatus;
  createdAt: string;
  updatedAt: string;
}

/** Payload for `POST /api/v1/applications` — starts an application for a job. */
export interface ApplicationCreateRequest {
  jobId: string;
  candidateProfileId: string;
  profileVersion: number | null;
  note: string | null;
}

export interface TransitionRequest {
  targetStatus: ApplicationStatus;
  note: string | null;
}

File: services/api/app/schemas/application.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.state_machine.application import ApplicationStatus


class ApplicationEventSchema(BaseModel):
    """Immutable audit event for a state transition."""

    id: str
    application_id: str
    from_status: str | None = None
    to_status: str
    triggered_by: str  # "system" | "human"
    actor_id: str | None = None
    note: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class PacketFingerprintSchema(BaseModel):
    """Captures the exact approved versions that constitute a submission packet."""

    profile_version: int
    resume_version: int | None = None
    cover_letter_version: int | None = None
    answer_versions: dict[str, int] = Field(default_factory=dict)
    job_snapshot_hash: str
    packet_hash: str


class ApplicationSchema(BaseModel):
    """Full application including audit event history."""

    id: str
    job_id: str
    candidate_profile_id: str
    status: str
    packet_fingerprint: PacketFingerprintSchema | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    events: list[ApplicationEventSchema] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class ApplicationSummarySchema(BaseModel):
    """Lightweight application for list views."""

    id: str
    job_id: str
    candidate_profile_id: str
    status: str
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class TransitionRequestSchema(BaseModel):
    """Request body for the POST /applications/{id}/transition endpoint."""

    target_status: ApplicationStatus
    note: str | None = None
